### PR TITLE
implement proposal for fetching paths and retaining structure of dirs and packages

### PR DESCRIPTION
**Pull request by madhur-tandon**
Closes https://github.com/pyscript/pyscript/issues/519 and https://github.com/pyscript/pyscript/issues/921

Proposal here: https://github.com/pyscript/pyscript/issues/519#issuecomment-1276276317

TODO:
- [x]  Update docs

### CODE REVIEW

File: docs/reference/elements/py-config.md [hunk 1]
```diff
@@ -75,7 +75,8 @@ One can also use both i.e pass the config from `src` attribute as well as specif
 
 ```html
 <py-config src="./custom.toml">
-  paths = ["./utils.py"]
+  [[fetch]]
+  files = ["./utils.py"]
 </py-config>
```
Comment (antocuni) [inline docs/reference/elements/py-config.md:80]: [suggestion] This example has the purpose of showing that you can use both `src` and inline content, it's probably better to use an easier-to-understand config settings, e.g. `packages` or `name`.

File: docs/reference/elements/py-config.md [hunk 2]
```diff
@@ -212,10 +217,17 @@ The following optional values are supported by `<py-config>`:
...
 | `plugins` | List of Plugins | List of Plugins are to be specified here. The default value is an empty list. |
 | `runtimes` | List of Runtimes | List of runtime configurations, described below. The default value contains a single Pyodide based runtime. |
 
+A fetch configuration consists of the following:
+| Value | Type | Description |
+| ----- | ---- | ----------- |
+| `from` | string | Base URL for the resource to be fetched. |
+| `to_folder` | string | Name of the folder to create in the filesystem. |
+| `files` | List of string | List of files to be downloaded. |
+
```
Comment (antocuni) [inline docs/reference/elements/py-config.md:231]: I think it's fine for this PR, but we should have better docs to explain how this works.
@tedpatrick do we have a process to signal "please write better docs for this topic" so that our technical writers know about it?
E.g. by putting `TODO: bla bla bla` comments directly inside the docs or creating an issue?

File: pyscriptjs/src/utils.ts [hunk 3]
```diff
@@ -109,3 +111,35 @@ export function getAttribute(el: Element, attr: string): string | null {
     }
     return null;
 }
+
+export function calculatePaths(fetch_cfg: FetchConfig[]) {
```
Comment (antocuni) [inline pyscriptjs/src/utils.ts:115]: why did yoy put this code into `utils.ts`?
It's specifically for `py-config`, it should go into `pyconfig.ts.`

File: pyscriptjs/src/utils.ts [hunk 4]
```diff
@@ -109,3 +111,35 @@ export function getAttribute(el: Element, attr: string): string | null {
...
+    const fetchPaths: string[] = [];
+    const paths: string[] = [];
+    fetch_cfg.forEach(function (each_fetch_cfg: FetchConfig) {
```
Comment (antocuni) [inline pyscriptjs/src/utils.ts:118]: [question] just my curiosity but: what is the advantage of using `forEach` instead of a normal `for (each_fetch_cfg of FetchConfig)`?

File: pyscriptjs/src/utils.ts [hunk 5]
```diff
@@ -109,3 +111,35 @@ export function getAttribute(el: Element, attr: string): string | null {
...
+        const from = each_fetch_cfg.from || "";
+        const to_folder = each_fetch_cfg.to_folder || ".";
+        const files = each_fetch_cfg.files;
```
Comment (antocuni) [inline pyscriptjs/src/utils.ts:122]: this is missing support for `to_file` as described here: https://github.com/pyscript/pyscript/issues/519#issuecomment-1302362956

File: pyscriptjs/src/utils.ts [hunk 6]
```diff
@@ -109,3 +111,35 @@ export function getAttribute(el: Element, attr: string): string | null {
...
+        if (files !== undefined)
+        {
+            for (const each_f of files)
+            {
+                const each_fetch_path = [from.endsWith('/') ? from.slice(0, -1) : from, each_f].filter(it => it!== "").join('/');
```
Comment (antocuni) [inline pyscriptjs/src/utils.ts:131]: I have no idea what this line does :). Sure, I could probably spend 5 minutes trying to understand it, OR you could write a nice function. If I understand correctly, there you want to do essentially the equivalent of `os.path.join`, don't you?
If so, I suggest to just write a small function `joinPaths` which does just this.
Bonus point: as soon as you have your own function, you can unit-test it very well, with all sort of corner cases.

Also: why do you have `filter(it => it !== "")`? Does it mean that if I have `files = ["a", "", "b"]`, the 2nd is silently ignored and it ends up fetching only 3 files?
If so, -1. It should be an error.

File: pyscriptjs/src/utils.ts [hunk 7]
```diff
@@ -109,3 +111,35 @@ export function getAttribute(el: Element, attr: string): string | null {
...
+        }
+        else
+        {
+            fetchPaths.push(from);
+            const filename = from.split('/').pop();
+            if (filename === '') {
+                throw Error(`Couldn't determine the filename from the path ${from}`);
+            }
+            else {
+                paths.push([to_folder.endsWith('/') ? to_folder.slice(0, -1) : to_folder, filename].join('/'));
```
Comment (antocuni) [inline pyscriptjs/src/utils.ts:145]: oh, and this is the 3rd time you copy&pasted the code. One more reason to make it a standalone function

File: pyscriptjs/src/utils.ts [hunk 8]
```diff
@@ -109,3 +111,35 @@ export function getAttribute(el: Element, attr: string): string | null {
...
+                const each_path = [to_folder.endsWith('/') ? to_folder.slice(0, -1) : to_folder, each_f].join('/');
+                paths.push(each_path);
+            }
+        }
+        else
+        {
+            fetchPaths.push(from);
+            const filename = from.split('/').pop();
+            if (filename === '') {
+                throw Error(`Couldn't determine the filename from the path ${from}`);
```
Comment (antocuni) [inline pyscriptjs/src/utils.ts]: small note: if PR #909 gets merged before this one, this shoud become `throw UserError`.

**Review by antocuni — CHANGES_REQUESTED** [review]
Good job, apart a couple of notes.
But, it completely misses integration tests which should to to `tests/integration/test_py_config.py`.

File: pyscriptjs/src/utils.ts (hunk 5 above)
Comment (madhur-tandon) [reply pyscriptjs/src/utils.ts:122]: @antocuni I think you reviewed a previous commit. I already added support for `to_file`. Not sure why you cannot see it.

File: docs/reference/elements/py-config.md (hunk 2 above)
Comment (madhur-tandon) [reply docs/reference/elements/py-config.md:231]: I guess you cannot see an entry for `to_file` here as well since I think you reviewed a previous commit.

File: docs/reference/elements/py-config.md (hunk 1 above)
Comment (madhur-tandon) [reply docs/reference/elements/py-config.md:80]: done, using `packages`

File: pyscriptjs/src/utils.ts (hunk 3 above)
Comment (madhur-tandon) [reply pyscriptjs/src/utils.ts:115]: put it in `plugins/fetch.ts` as discussed on slack

File: pyscriptjs/src/utils.ts (hunk 7 above)
Comment (madhur-tandon) [reply pyscriptjs/src/utils.ts:145]: done

File: pyscriptjs/src/utils.ts (hunk 6 above)
Comment (madhur-tandon) [reply pyscriptjs/src/utils.ts:131]: done, added testing, the part about `filter(it => it !== "")` is needed since the behaviour of
`os.path.join('', '', '')` is simply `''` so we are replicating that in some sense.

File: pyscriptjs/src/utils.ts (hunk 5 above)
Comment (madhur-tandon) [reply pyscriptjs/src/utils.ts:122]: resolving...

**Comment by madhur-tandon** [comment]
@antocuni For the integration tests, they were present in `test_01_basic.py`, I have moved it to `test_py_config.py`.
Let me know if you have ideas for more of them.

I am not sure if we should write tests using a custom URL (eg: `http://example.com/data.csv`) -- since:
- I don't know such example websites that host some dummy files
- Do we want tests that depend on network?

File: pyscriptjs/src/utils.ts (hunk 4 above)
Comment (madhur-tandon) [reply pyscriptjs/src/utils.ts]: I don't know to be honest except that it is considered a standard practice I guess? If not, I can change. Let me know.

Comment (antocuni) [reply pyscriptjs/src/utils.ts]: I don't really know because I'm not an expert of JS/TS.
In all the rest of the code we use `for(... of ...)`, that's why I wondered whether here it was different

**Comment by antocuni** [comment]
> I am not sure if we should write tests using a custom URL (eg: `http://example.com/data.csv`) -- since:

if this case is already covered by unit tests, I think we are fine

File: pyscriptjs/tests/unit/utils.test.ts [hunk 9]
```diff
@@ -34,49 +33,22 @@ describe("Utils", () => {
...
-  it("should calculate paths when only from is provided", () => {
-    const fetch_cfg: FetchConfig[] = [{from: "http://a.com/data.csv" }];
-    const [paths, fetchPaths] = calculatePaths(fetch_cfg);
-    expect(paths).toStrictEqual(["./data.csv"]);
-    expect(fetchPaths).toStrictEqual(["http://a.com/data.csv"]);
+describe("JoinPaths", () => {
+  it("should remove trailing slashes from the beginning and the end", () => {
+    const paths: string[] = ['///abc/d/e///'];
+    const joinedPath = joinPaths(paths);
+    expect(joinedPath).toStrictEqual('abc/d/e');
```
Comment (antocuni) [inline pyscriptjs/tests/unit/utils.test.ts:40]: this looks wrong.
If I do `jointPaths("/tmp", "foo.txt")` I expect `/tmp/foo.txt`, not `tmp/foo.txt`

**Review by antocuni — APPROVED** [review]
good job :muscle: 
I'm approving it in advance so that you can merge it without any further review from me, but the logic about "remove leading slashes" should be fixed before merging

File: pyscriptjs/tests/unit/utils.test.ts (hunk 9 above)
Comment (madhur-tandon) [reply pyscriptjs/tests/unit/utils.test.ts:40]: Done, I only check for a leading slash on the first element though. Otherwise it will be very cumbersome.

File: pyscriptjs/src/utils.ts (hunk 8 above)
Comment (madhur-tandon) [reply pyscriptjs/src/utils.ts]: Done in #927
